### PR TITLE
SQSERVICES-1028-be-add-sso-scim-attributes-to-user-profile-for-team-admins

### PR DESCRIPTION
**Pull request by battermann**
https://wearezeta.atlassian.net/browse/SQSERVICES-1028

## Checklist

 - [x] The **PR Title** explains the impact of the change.
 - [x] The **PR description** provides context as to why the change should occur and what the code contributes to that effect. This could also be a link to a JIRA ticket or a Github issue, if there is one.
 - [x] **changelog.d** contains the following bits of information ([details](https://github.com/wireapp/wire-server/blob/develop/docs/developer/changelog.md)):
   - [x] A file with the changelog entry in one or more suitable sub-sections. The sub-sections are marked by directories inside `changelog.d`.

### CODE REVIEW

File: libs/wire-api/src/Wire/API/User/Search.hs [hunk 1]
```diff
@@ -200,6 +223,9 @@ modelTeamContact = Doc.defineModel "TeamContact" $ do
   Doc.property "email" Doc.string' $ do
     Doc.description "Email address"
     Doc.optional
+  Doc.property "scim_external_id" Doc.string' $ do
```
Comment (fisx) [inline libs/wire-api/src/Wire/API/User/Search.hs:226]: what about `sso`?

File: libs/wire-api/src/Wire/API/User/Search.hs [hunk 2]
```diff
@@ -166,6 +167,26 @@ instance ToSchema Contact where
 --------------------------------------------------------------------------------
 -- TeamContact
 
+data Sso = Sso
```
Comment (fisx) [inline libs/wire-api/src/Wire/API/User/Search.hs:170]:
```suggestion
-- | Related to `UserSSOId`, but more straight-forward because it does not take SCIM externalId into account.
data Sso = Sso
```

... or something to that end?

File: services/brig/src/Brig/User/Search/Index.hs [hunk 3]
```diff
@@ -586,7 +588,38 @@ indexMapping =
                   mpIndex = True,
                   mpAnalyzer = Nothing,
                   mpFields = mempty
-                }
+                },
+            "scim_external_id"
+              .= MappingProperty
+                { mpType = MPKeyword,
+                  mpStore = False,
+                  mpIndex = False,
```
Comment (fisx) [inline services/brig/src/Brig/User/Search/Index.hs:596]: On second thought, can you double-check if this is something we need for implementing the current designs?  Would be annoying if we had to make another PR, another migration...  (sorry for the back & forth, but it's better not to index things we don't need indices for, ES takes a lot of resources).

File: services/brig/src/Brig/User/Search/Index.hs [hunk 4]
```diff
@@ -827,9 +862,17 @@ reindexRowToIndexUser
...
-        Just $ (cs . toLazyByteString . serializeURIRef) uri
+        Just $ fromUri uri
       idpUrl (UserScimExternalId _) = Nothing
 
+      fromUri :: URI -> Text
+      fromUri = cs . toLazyByteString . serializeURIRef
+
+      sso :: UserSSOId -> Maybe Sso
+      sso userSsoId = do
+        (issuer, nameid) <- User.ssoIssuerAndNameId userSsoId
```
Comment (fisx) [inline services/brig/src/Brig/User/Search/Index.hs:873]: [optional] Why the detour through the tuple?  I would have named `Sso` `SsoOnly` (or something else more specific) and typed `User.ssoIssuerAndNameId :: UserSSOId -> Maybe SsoOnly`.

File: services/brig/src/Brig/User/Search/Index/Types.hs [hunk 5]
```diff
@@ -60,7 +61,9 @@ data IndexUser = IndexUser
     _iuManagedBy :: Maybe ManagedBy,
     _iuCreatedAt :: Maybe UTCTime,
     _iuRole :: Maybe Role,
-    _iuSearchVisibilityInbound :: Maybe SearchVisibilityInbound
+    _iuSearchVisibilityInbound :: Maybe SearchVisibilityInbound,
+    _iuScimExternalId :: Maybe Text,
+    _iuSso :: Maybe Sso
```
Comment (fisx) [inline services/brig/src/Brig/User/Search/Index/Types.hs:66]: This may be correct (I don't know ES), but are you sure if a field is not indexed, it should still appear here and not just in `UserDoc`?  If so, why?

**Review by fisx — APPROVED** [review]

File: services/brig/src/Brig/User/Search/Index.hs (hunk 4 above)
Comment (battermann) [reply services/brig/src/Brig/User/Search/Index.hs:873]: Yes, but this would make a cyclic dependency, therefore the tuple.

File: services/brig/src/Brig/User/Search/Index/Types.hs (hunk 5 above)
Comment (battermann) [reply services/brig/src/Brig/User/Search/Index/Types.hs:66]: AFAIK `UserDoc` is a subset of the index.

File: services/brig/src/Brig/User/Search/Index.hs (hunk 3 above)
Comment (battermann) [reply services/brig/src/Brig/User/Search/Index.hs:596]: Not sure, I understand the concern. The field is explicitly not indexed.
